File: examples/modulith-clinical-tool-jpa/src/main/java/io/zenwave360/example/clinicaltool/modules/masterdata/MasterDataApiController.java
package io.zenwave360.example.clinicaltool.modules.masterdata;

import io.zenwave360.example.clinicaltool.modules.masterdata.domain.*;
import io.zenwave360.example.clinicaltool.modules.masterdata.dtos.*;
import io.zenwave360.example.clinicaltool.modules.masterdata.mappers.*;
import java.math.*;
import java.time.*;
import java.util.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.context.request.NativeWebRequest;

/**
 * REST controller for MasterDataApi.
 */
@RestController
@RequestMapping("/api")
public class MasterDataApiController implements MasterDataApi {

    private final Logger log = LoggerFactory.getLogger(getClass());

    @Autowired
    private NativeWebRequest request;

    private MasterDataService masterDataService;

    @Autowired
    public MasterDataApiController setMasterDataService(MasterDataService masterDataService) {
        this.masterDataService = masterDataService;
        return this;
    }

    private MasterDataDTOsMapper mapper = MasterDataDTOsMapper.INSTANCE;

    public MasterDataApiController(MasterDataService masterDataService) {

        this.masterDataService = masterDataService;
    }

    public Optional<NativeWebRequest> getRequest() {
        return Optional.ofNullable(request);
    }

    @Override
    public ResponseEntity<MasterDataDTO> createMasterData(MasterDataDTO reqBody) {
        log.debug("REST request to createMasterData: {}", reqBody);
        var input = mapper.asMasterData(reqBody);
        var masterData = masterDataService.createMasterData(input);
        MasterDataDTO responseDTO = mapper.asMasterDataDTO(masterData);
        return ResponseEntity.status(201).body(responseDTO);
    }

    @Override
    public ResponseEntity<MasterDataPaginatedDTO> listMasterData(Integer page, Integer limit, List<String> sort) {
        log.debug("REST request to listMasterData: {}, {}, {}", page, limit, sort);
        var masterDataPage = masterDataService.listMasterData(pageOf(page, limit, sort));
        var responseDTO = mapper.asMasterDataPaginatedDTO(masterDataPage);
        return ResponseEntity.status(200).body(responseDTO);
    }

    @Override
    public ResponseEntity<MasterDataDTO> getMasterData(Long id) {
        log.debug("REST request to getMasterData: {}", id);
        var masterData = masterDataService.getMasterData(id);
        if (masterData.isPresent()) {
            MasterDataDTO responseDTO = mapper.asMasterDataDTO(masterData.get());
            return ResponseEntity.status(200).body(responseDTO);
        } else {
            return ResponseEntity.notFound().build();
        }
    }

    @Override
    public ResponseEntity<MasterDataDTO> updateMasterData(Long id, MasterDataDTO reqBody) {
        log.debug("REST request to updateMasterData: {}, {}", id, reqBody);
        var input = mapper.asMasterData(reqBody);
        var masterData = masterDataService.updateMasterData(id, input);
        if (masterData.isPresent()) {
            MasterDataDTO responseDTO = mapper.asMasterDataDTO(masterData.get());
            return ResponseEntity.status(200).body(responseDTO);
        } else {
            return ResponseEntity.notFound().build();
        }
    }

    @Override
    public ResponseEntity<Void> deleteMasterData(Long id) {
        log.debug("REST request to deleteMasterData: {}", id);
        masterDataService.deleteMasterData(id);
        return ResponseEntity.status(204).build();
    }

    @Override
    public ResponseEntity<List<MasterDataKeyValueDTO>> listMasterDataOfType(String type, String lang) {
        log.debug("REST request to listMasterDataOfType: {}, {}", type, lang);
        var masterDataKeyValue = masterDataService.listMasterDataOfType(MasterDataType.valueOf(type), lang);
        var responseDTO = mapper.asMasterDataKeyValueDTOList(masterDataKeyValue);
        return ResponseEntity.status(200).body(responseDTO);
    }

    protected Pageable pageOf(Integer page, Integer limit, List<String> sort) {
        Sort sortOrder = sort != null
                ? Sort.by(sort.stream()
                        .map(sortParam -> {
                            String[] parts = sortParam.split(":");
                            String property = parts[0];
                            Sort.Direction direction =
                                    parts.length > 1 ? Sort.Direction.fromString(parts[1]) : Sort.Direction.ASC;
                            return new Sort.Order(direction, property);
                        })
                        .toList())
                : Sort.unsorted();
        return PageRequest.of(page != null ? page : 0, limit != null ? limit : 10, sortOrder);
    }
}

File: examples/modulith-klinical-tool-jpa/src/main/kotlin/io/zenwave360/example/clinicaltool/modules/masterdata/MasterDataApiController.kt
package io.zenwave360.example.clinicaltool.modules.masterdata

import io.zenwave360.example.clinicaltool.modules.masterdata.domain.*
import io.zenwave360.example.clinicaltool.modules.masterdata.*
import io.zenwave360.example.clinicaltool.modules.masterdata.dtos.*
import io.zenwave360.example.clinicaltool.modules.masterdata.*
import io.zenwave360.example.clinicaltool.modules.masterdata.dtos.*
import io.zenwave360.example.clinicaltool.common.*
import io.zenwave360.example.clinicaltool.modules.masterdata.mappers.*

import java.net.URI
import java.net.URISyntaxException
import java.math.*
import java.time.*
import java.util.*
import jakarta.validation.Valid
import jakarta.validation.constraints.NotNull
import org.mapstruct.factory.Mappers
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.annotation.Value
import org.springframework.http.MediaType
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import org.springframework.core.io.ByteArrayResource
import org.springframework.core.io.Resource
import org.springframework.data.domain.Page
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Pageable
import org.springframework.data.domain.Sort
import org.springframework.web.context.request.NativeWebRequest

/**
 * REST controller for MasterDataApi.
 */
@RestController
@RequestMapping("/api")
open class MasterDataApiController(
    private val masterDataService: MasterDataService
) : MasterDataApi {

    private val log: Logger = LoggerFactory.getLogger(javaClass)

    @Autowired
    private lateinit var request: NativeWebRequest

    private val mapper = MasterDataDTOsMapper.INSTANCE



    override fun createMasterData(reqBody: MasterDataDTO): ResponseEntity<MasterDataDTO> {
        log.debug("REST request to createMasterData: {}", reqBody)
        val input = mapper.asMasterData(reqBody)
        val masterData =  masterDataService.createMasterData(input)
        val responseDTO = mapper.asMasterDataDTO(masterData)
        return ResponseEntity.status(201).body(responseDTO)
    }

    override fun listMasterData(page: Int, limit: Int, sort: List<String>?): ResponseEntity<MasterDataPaginatedDTO> {
        log.debug("REST request to listMasterData: {}, {}, {}", page, limit, sort)
        val masterDataPage =  masterDataService.listMasterData(pageOf(page, limit, sort))
        val responseDTO = mapper.asMasterDataPaginatedDTO(masterDataPage)
        return ResponseEntity.status(200).body(responseDTO)
    }

    override fun getMasterData(id: Long): ResponseEntity<MasterDataDTO> {
        log.debug("REST request to getMasterData: {}", id)
        val masterData =  masterDataService.getMasterData(id)
        return if (masterData != null) {
            val responseDTO = mapper.asMasterDataDTO(masterData)
            ResponseEntity.status(200).body(responseDTO)
        } else {
            ResponseEntity.notFound().build()
        }
    }

    override fun updateMasterData(id: Long, reqBody: MasterDataDTO): ResponseEntity<MasterDataDTO> {
        log.debug("REST request to updateMasterData: {}, {}", id, reqBody)
        val input = mapper.asMasterData(reqBody)
        val masterData =  masterDataService.updateMasterData(id, input)
        return if (masterData != null) {
            val responseDTO = mapper.asMasterDataDTO(masterData)
            ResponseEntity.status(200).body(responseDTO)
        } else {
            ResponseEntity.notFound().build()
        }
    }

    override fun deleteMasterData(id: Long): ResponseEntity<Unit> {
        log.debug("REST request to deleteMasterData: {}", id)
        masterDataService.deleteMasterData(id)
        return ResponseEntity.status(204).build()
    }

    override fun listMasterDataOfType(type: String, lang: String): ResponseEntity<List<MasterDataKeyValueDTO>> {
        log.debug("REST request to listMasterDataOfType: {}, {}", type, lang)
        val masterDataKeyValue =  masterDataService.listMasterDataOfType(MasterDataType.valueOf(type), lang)
        val responseDTO = mapper.asMasterDataKeyValueDTOList(masterDataKeyValue)
        return ResponseEntity.status(200).body(responseDTO)
    }


    protected fun pageOf(page: Int?, limit: Int?, sort: List<String>?): Pageable {
        val sortOrder = sort?.let {
            Sort.by(it.map { sortParam ->
                val parts = sortParam.split(":")
                val property = parts[0]
                val direction = if (parts.size > 1) Sort.Direction.fromString(parts[1]) else Sort.Direction.ASC
                Sort.Order(direction, property)
            })
        } ?: Sort.unsorted()
        return PageRequest.of(page ?: 0, limit ?: 10, sortOrder)
    }
}
